provider/provider.go

```diff
@@ -692,7 +725,35 @@ func (s *SweepingProvider) getAvgPrefixLenNoLock() (int, error) {
 		s.cachedAvgPrefixLen = prefixLenSum / len(scheduleEntries)
 		s.lastAvgPrefixLen = time.Now()
 	}
-	return s.cachedAvgPrefixLen, nil
+	return s.cachedAvgPrefixLen
+}
+
+func (s *SweepingProvider) persistAvgPrefixLen() error {
+	l := s.getAvgPrefixLenNoLock()
+
+	if l < 0 {
+		// Node is offline, don't persit prefix length.
+		return nil
+	}
+	if l > keyspace.KeyLen {
+		return fmt.Errorf("cannot persist invalid average prefix length %d", l)
+	}
+	return s.datastore.Put(s.ctx, avgPrefixLenDatastoreKey, []byte{byte(l)})
+}
+
+func loadAvgPrefixLen(ctx context.Context, ds datastore.Batching) (int, error) {
+	val, err := ds.Get(ctx, avgPrefixLenDatastoreKey)
+	if err != nil {
+		if errors.Is(err, datastore.ErrNotFound) {
+			// Suppress error if not found, -1 indicates missing value.
+			err = nil
+		}
+		return -1, err
+	}
+	if len(val) != 1 {
+		return -1, fmt.Errorf("invalid avg prefix len value in datastore")
+	}
+	return int(val[0]), nil
```

Comment:
nit: while `persistAvgPrefixLen` validates `l <= keyspace.KeyLen` before persisting, `loadAvgPrefixLen` doesn't validate on load.  Could we add extra check on load too?

Rationale: If the datastore is corrupted or there's a bug, this could return an invalid value.  It sound like theoretical risk, but remember 60+% of IPFS Desktop users run on windows, and filesystem there is... let's say we did have disproportionate number of JSON file corruptions.



```go
l := int(val[0])
if l > keyspace.KeyLen {
    return -1, fmt.Errorf("invalid avg prefix len in datastore: %d > %d", l, keyspace.KeyLen)
}
return l, nil
```

Reply:
Added check in https://github.com/libp2p/go-libp2p-kad-dht/pull/1176/commits/30046208be4513f09474dfd574f1bcb58eb4f041.

It shouldn't be necessary, since the number is stored on a single byte, (`keyspace.KeyLen = 256`). If stored value isn't exactly 1 byte long, we error anyway. So the range is strictly limited between 0 and 255 (`keyspace.KeyLen-1`).

Defensive programming is always safer though.